std/experimental/allocator/package.d

```diff
@@ -1123,48 +1123,133 @@ allocated with the same allocator.
 */
 void dispose(A, T)(auto ref A alloc, T* p)
 {
+    if (!p) return;
     static if (hasElaborateDestructor!T)
-    {
-        destroy(*p);
-    }
+        p.__dtor;
     alloc.deallocate((cast(void*)p)[0 .. T.sizeof]);
 }
 
 /// Ditto
 void dispose(A, T)(auto ref A alloc, T p)
-if (is(T == class) || is(T == interface))
+if (is(T == class) && T.stringof != Object.stringof)
 {
     if (!p) return;
-    static if (is(T == interface))
+    auto support = (cast(void*) p)[0 .. typeid(p).initializer.length];
+    static if (__traits(hasMember, T, "__dtor"))
+        p.__dtor;
+    else
```

Comment:
Right, I've removed the redundant code. I've also modified the code so that dispose is guaranteed to be always inlined (even by DMD).